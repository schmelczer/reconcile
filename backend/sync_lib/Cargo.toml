[package]
name = "sync_lib"
version.workspace = true
edition.workspace = true
authors.workspace = true
license.workspace = true
repository.workspace = true

[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
base64 = "0.22.1"
reconcile = { path = "../reconcile" }
wasm-bindgen = "0.2.99"
thiserror = { workspace = true }

# The `console_error_panic_hook` crate provides better debugging of panics by
# logging them with `console.error`. This is great for development, but requires
# all the `std::fmt` and `std::panicking` infrastructure, so isn't great for
# code size when deploying.
console_error_panic_hook = { version = "0.1.7", optional = true }

[dev-dependencies]
wasm-bindgen-test = "0.3.49"
insta = "1.41.1"

[features]
default = ["console_error_panic_hook"]

[lints]
workspace = true
