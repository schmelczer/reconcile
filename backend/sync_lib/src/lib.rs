use core::str;

use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use errors::SyncLibError;
use wasm_bindgen::prelude::*;

pub mod errors;

// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
// allocator.
#[cfg(feature = "wee_alloc")]
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

#[wasm_bindgen]
pub fn bytes_to_base64(input: &[u8]) -> String { STANDARD_NO_PAD.encode(input) }

#[wasm_bindgen]
pub fn string_to_base64(input: &str) -> String { bytes_to_base64(input.as_bytes()) }

#[wasm_bindgen]
pub fn base64_to_bytes(input: &str) -> Result<Vec<u8>, SyncLibError> {
    STANDARD_NO_PAD.decode(input).map_err(SyncLibError::from)
}

#[wasm_bindgen]
pub fn base64_to_string(input: &str) -> Result<String, SyncLibError> {
    let bytes = base64_to_bytes(input)?;
    String::from_utf8(bytes).map_err(SyncLibError::from)
}

#[wasm_bindgen]
pub fn merge(parent: &[u8], left: &[u8], right: &[u8]) -> Result<Vec<u8>, SyncLibError> {
    Ok(if is_binary(right) {
        right.to_vec()
    } else {
        reconcile::reconcile(
            str::from_utf8(parent).map_err(SyncLibError::from)?,
            str::from_utf8(left).map_err(SyncLibError::from)?,
            str::from_utf8(right).map_err(SyncLibError::from)?,
        )
        .into_bytes()
    })
}

#[wasm_bindgen]
pub fn is_binary(data: &[u8]) -> bool { data.iter().any(|&b| b == 0) }

pub fn set_panic_hook() {
    // When the `console_error_panic_hook` feature is enabled, we can call the
    // `set_panic_hook` function at least once during initialization, and then
    // we will get better error messages if our code ever panics.
    //
    // For more details see
    // https://github.com/rustwasm/console_error_panic_hook#readme
    #[cfg(feature = "console_error_panic_hook")]
    console_error_panic_hook::set_once();
}
