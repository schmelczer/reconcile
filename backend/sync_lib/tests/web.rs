//! Test suite for the Web and headless browsers.

use insta::assert_debug_snapshot;
use sync_lib::*;
use wasm_bindgen_test::*;

#[wasm_bindgen_test(unsupported = test)]
fn test_bytes_to_base64() {
    let input = b"hello";
    let expected = "aGVsbG8";
    assert_eq!(bytes_to_base64(input), expected);
}

#[wasm_bindgen_test(unsupported = test)]
fn test_base64_to_bytes() {
    let input = "aGVsbG8";
    let expected = b"hello".to_vec();
    assert_eq!(base64_to_bytes(input).unwrap(), expected);
}

#[test] // insta doesn't support wasm-bindgen-test
fn test_base64_to_bytes_error() {
    let input = "===";
    assert_debug_snapshot!(base64_to_bytes(input));
}

#[wasm_bindgen_test(unsupported = test)]
fn merge_text() {
    let left = b"hello ";
    let right = b"world";
    assert_eq!(merge(b"", left, right), b"hello world".to_vec());
}

#[wasm_bindgen_test(unsupported = test)]
fn merge_binary() {
    let left = [0, 1, 2];
    let right = [3, 4, 5];
    assert_eq!(merge(b"", &left, &right), right);
}

#[wasm_bindgen_test(unsupported = test)]
fn test_is_binary() {
    assert!(is_binary(&[0, 159, 146, 150]));
    assert!(is_binary(&[0, 12]));
    assert!(!is_binary(b"hello"));
}
