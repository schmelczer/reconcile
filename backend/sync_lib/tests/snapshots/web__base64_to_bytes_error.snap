---
source: sync_lib/tests/web.rs
expression: base64_to_bytes(input)
snapshot_kind: text
---
Err(
    Base64DecodingError {
        reason: "Invalid symbol 61, offset 0.",
    },
)
